### PR TITLE
Attachments server

### PR DESCRIPTION
**Pull request by zwade**
Since the whole attachments pull request is so big, I went ahead and broke it into two parts, server and client. Here is the first part, which is all of the backend requirements. As of yet, this does not contain any additional testing, however it does pass all of the existing tests. Will update with testing once I determine the best way to include them.

### CODE REVIEW

File: scoreboard/models.py [hunk 1]
```diff
@@ -552,16 +551,21 @@ def update_answers(self, exclude_team=None):
                 ScoreHistory.add_entry(a.team)
 
 
+attch_challenge_association = db.Table('attch_chall_association', db.Model.metadata,
```
Comment (k9delight) [inline scoreboard/models.py]: any particular reason there is an "a" missing from the variable name?

File: scoreboard/models.py [hunk 2]
```diff
@@ -552,16 +551,21 @@ def update_answers(self, exclude_team=None):
...
 
     aid = db.Column(db.String(64), primary_key=True)
-    challenge_cid = db.Column(db.BigInteger, db.ForeignKey('challenge.cid'),
-            nullable=False)
     filename = db.Column(db.String(100), nullable=False)
     content_type = db.Column(db.String(100))
     storage_path = db.Column(db.String(256))
 
+    challenges = db.relationship('Challenge', backref=db.backref('attachments', lazy='joined'),
+                                 secondary='attch_chall_association', lazy='joined')
```
Comment (k9delight) [inline scoreboard/models.py]: might need a s/old/new/ variable name :)

**Comment by zwade** [comment]
Finally got tests working, will address other issues tomorrow

File: scoreboard/rest.py [hunk 3]
```diff
@@ -801,18 +801,76 @@ def delete(self, path):
...
+            models.commit()
+            cache.clear()
         return dict(aid=aid, fpath=fpath, content_type=fp.mimetype)
 
-api.add_resource(Upload, '/api/upload')
+    @flask_restful.marshal_with(resource_fields)
+    def get(self):
+        return dict(attachments=list(models.Attachment.query.all()))
+
+api.add_resource(Attachment, '/api/attachment/<string:aid>')
```
Comment (Matir) [inline scoreboard/rest.py]: For consistency, can we make this /api/attachments?

File: scoreboard/attachments/__init__.py [hunk 4]
```diff
@@ -40,11 +41,23 @@ def get_backend_type():
     backend = get_backend_path()
     return urlparse.urlparse(backend).scheme
 
+def patch(_backend_type):
```
Comment (Matir) [inline scoreboard/attachments/__init__.py]: This feels very repetitive.  We could have a method (such as "get_backend"), that does the comparison/import, then returns `backend`?  This way we don't have to repeat the if statements/imports

File: scoreboard/attachments/__init__.py [hunk 5]
```diff
@@ -40,11 +41,23 @@ def get_backend_type():
...
+    elif _backend_type == "test":
+        import testing as backend
+    else:
+        raise ImportError('Unhandled attachment backend %s' % _backend_type)
+    globals()['backend'] = backend
 
 _backend_type = get_backend_type()
 if _backend_type == "file":
-    from scoreboard.attachments.file import *
+    import file as backend
```
Comment (Matir) [inline scoreboard/attachments/__init__.py]: Is there a reason the import style had to change?

Specifically, I get you wanted to import the module as backend, but `from scoreboard.attachments import file as backend` achieves that.  

Python 3 has actually removed this implicit relative import, so I'd like to keep the full package paths where possible.  (Yes, we need to fix a lot of other things for py3k support, but I'd rather not break new things.)

File: scoreboard/models.py [hunk 6]
```diff
@@ -507,11 +505,12 @@ def set_attachments(self, attachments):
             aid_set.add(a['aid'])
             attachment = Attachment.query.get(a['aid'])
             if not attachment:
-                Attachment.create(a['aid'], a['filename'], a['content_type'], self)
+                logging.warning('Trying to add attachment that does not exist')
```
Comment (Matir) [inline scoreboard/models.py]: Let's log the attachment id as well, as that will help debugging issues that may occur.

File: scoreboard/models.py [hunk 7]
```diff
@@ -552,16 +551,21 @@ def update_answers(self, exclude_team=None):
                 ScoreHistory.add_entry(a.team)
 
 
+attch_challenge_association = db.Table('attch_chall_association', db.Model.metadata,
+        db.Column('challenge_cid', db.BigInteger,  db.ForeignKey('challenge.cid')),
+        db.Column('attachmemt_aid',   db.String(64), db.ForeignKey('attachment.aid')))
```
Comment (Matir) [inline scoreboard/models.py]: Do you mean attachment_aid?

File: scoreboard/models.py (hunk 1 above)
Comment (zwade) [inline scoreboard/models.py]: Not a significant one

File: scoreboard/attachments/__init__.py (hunk 5 above)
Comment (zwade) [inline scoreboard/attachments/__init__.py]: Ahh! That's what I was looking for. Sorry, python imports never cease to confuse me

**Pull request merged**